test/infrastructure/vcsim/controllers/envvar_controller.go

```diff
@@ -287,14 +293,74 @@ func clusterEnvVarSpecGovmomiVariables(c *vcsimv1.ClusterEnvVarSpec) map[string]
 	}
 
 	// NOTE: omitting cluster Name intentionally because E2E tests provide this value in other ways
-	vars["VSPHERE_DATACENTER"] = vcsimhelpers.DatacenterName(datacenter)
+	// Vars below can be named only so they should be used as is
 	vars["VSPHERE_COMPUTE_CLUSTER"] = vcsimhelpers.ClusterName(datacenter, cluster)
-	vars["VSPHERE_DATASTORE"] = vcsimhelpers.DatastoreName(datastore)
-	vars["VSPHERE_FOLDER"] = vcsimhelpers.VMFolderName(datacenter)
-	vars["VSPHERE_NETWORK"] = vcsimhelpers.NetworkPath(datacenter, vcsimhelpers.DefaultNetworkName)
-	vars["VSPHERE_RESOURCE_POOL"] = vcsimhelpers.ResourcePoolPath(datacenter, cluster)
-	vars["VSPHERE_TEMPLATE"] = vcsimhelpers.VMPath(datacenter, template)
-	return vars
+
+	if !moid {
+		vars["VSPHERE_DATACENTER"] = vcsimhelpers.DatacenterName(datacenter)
+		vars["VSPHERE_DATASTORE"] = vcsimhelpers.DatastoreName(datastore)
+		vars["VSPHERE_FOLDER"] = vcsimhelpers.VMFolderName(datacenter)
+		vars["VSPHERE_RESOURCE_POOL"] = vcsimhelpers.ResourcePoolPath(datacenter, cluster)
+		vars["VSPHERE_NETWORK"] = vcsimhelpers.NetworkPath(datacenter, vcsimhelpers.DefaultNetworkName)
+		vars["VSPHERE_TEMPLATE"] = vcsimhelpers.VMPath(datacenter, template)
+
+		return vars, nil
+	}
+
+	params := session.NewParams().
+		WithServer(existing["VSPHERE_SERVER"]).
+		WithThumbprint(existing["VSPHERE_TLS_THUMBPRINT"]).
+		WithUserInfo(existing["VSPHERE_USERNAME"], existing["VSPHERE_PASSWORD"])
+
+	s, err := session.GetOrCreate(ctx, params)
+	if err != nil {
+		return vars, err
+	}
+
+	var dc *object.Datacenter
+	dcRef := object.ReferenceFromString(vcsimhelpers.DatacenterName(datacenter))
+	if dcRef != nil {
+		dc = object.NewDatacenter(s.Client.Client, *dcRef)
+	} else {
+		dc, err = s.Finder.Datacenter(ctx, vcsimhelpers.DatacenterName(datacenter))
+		if err != nil {
+			return vars, fmt.Errorf("failed to locate datacenter reference: %w", err)
+		}
+	}
+	vars["VSPHERE_DATACENTER"] = dc.Reference().String()
+	s.Finder.SetDatacenter(dc)
+
+	dsRef, err := s.Finder.Datastore(ctx, vcsimhelpers.DatastoreName(datastore))
+	if err != nil {
+		return vars, fmt.Errorf("failed to locate datastore reference: %w", err)
```

Comment:
for consistency, please use pkg/errors with %v instead of introducing fmt.Error with %w
applies to more places below...